### ISSUE TITLE
Weird behavior on Prediction using loading the model using Keras load_model()

### ISSUE THREAD

**Issue by debhazra** [keras-team/keras#9279]
I have trained my model using EMNIST byclass dataset by loading .csv files of training and testing data as well as label for digit (0-9) and letter (A-Z, a-z) classification. I got the model evaluation accuracy around 87%. But when I am using the the best saved model weights (.hdf5) using keras load_model(), it gives me weird results, as if no training has been done.
****
I have not done any further model compilation before I used keras load_model() and model.predict_classes().

Only one point just FYIP and most likely it doesn't create any problem - I have loaded training and testing data from EMNIST .csv class file where as I am trying to predict the image which is in .jpg format (not through .csv file).

Can anyone please help me to figure out the solution and why such issue may be happening ?

Thanks
- Deb

**Comment by brainnoise** [comment]
Can you please share standalone code snippet to help reproduce the issue?

**Comment by debhazra** [comment]
`import pandas as pd
import numpy as np
np.random.seed(1337) # for reproducibility

from keras import backend as K
from keras.models import Sequential
from keras.layers.core import Dense, Dropout, Activation, Flatten
from keras.layers.convolutional import Convolution2D, MaxPooling2D
from keras.callbacks import Callback, RemoteMonitor
from keras.utils import np_utils

train = pd.read_csv("emnist/emnist-byclass-train.csv").values
test  = pd.read_csv("emnist/emnist-byclass-test.csv").values

print('train shape:', train.shape)
print('test shape:', test.shape)

img_rows = 28
img_cols = 28
X_train = train[:, 1:].reshape(train.shape[0], img_rows, img_cols, 1)
X_test = test[:, 1:].reshape(test.shape[0], img_rows, img_cols, 1)

y_train = train[:, 0] # First data is label (already removed from X_train)
y_test = test[:, 0] # First data is label (already removed from Y_train)

# Make the value floats in [0;1] instead of int in [0;255]
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255
X_test /= 255

# convert class vectors to binary class matrices (ie one-hot vectors)
Y_train = np_utils.to_categorical(y_train, nb_classes)
Y_test = np_utils.to_categorical(y_test, nb_classes)

### Define Model Architecture

model = Sequential()
 
model.add(Conv2D(filters=16, kernel_size=3, activation='relu', input_shape=(28, 28, 1)))
model.add(Conv2D(filters=32, kernel_size=3, activation='relu'))
model.add(MaxPooling2D(pool_size=(2,2)))
model.add(Conv2D(filters=64, kernel_size=3, activation='relu'))
#model.add(MaxPooling2D(pool_size=(2,2)))
#model.add(Dropout(0.25))
 
#model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(GlobalAveragePooling2D())
model.add(Dropout(0.25))
model.add(Dense(62, activation='softmax'))

model.summary()

from keras import optimizers
#rmsprop = optimizers.RMSprop(lr=1.0e-4)
Adamax = optimizers.Adamax(lr=0.00002)
#model.compile(optimizer='rmsprop', loss='categorical_crossentropy', metrics=['accuracy'])
model.compile(optimizer='Adamax', loss='categorical_crossentropy', metrics=['accuracy'])

from keras.callbacks import ModelCheckpoint 

epochs = 50
batch_size = 128

### Do NOT modify the code below this line.

checkpointer = ModelCheckpoint(filepath='saved_models/local_machine_DigitAlpha_EMNIST_best_model.hdf5', 
                               verbose=1, save_best_only=True)

noaugmented_history_epoch = model.fit(X_train, Y_train, 
          validation_data=(X_test, Y_test),
          epochs=epochs, batch_size=batch_size, callbacks=[checkpointer], verbose=1)

model.load_weights('saved_models/best_model.hdf5')

# 10. Evaluate model on test data
score = model.evaluate(X_test, Y_test, verbose = 0)

print("Test Accuracy with first 50 epochs: %.2f%%" % (score[1]*100))

# Predict unseen image
img_path = 'images/20180203_082156.jpg'
image = cv2.imread(img_path)
gray = cv2.cvtColor(image,cv2.COLOR_BGR2GRAY)
gray = cv2.resize(gray,(28,28))

gray = gray.reshape(1,28,28,1)
gray = gray.astype('float32')
gray /= 255

# Load the model and predict

model = 'saved_models/best_model.hdf5'
model = load_model(model)

label = model.predict_classes(gray)
display(label)`

**Comment by debhazra** [comment]
As you have requested, I have given the code snippet. Do you find any solution ? I am struggling with this issue since one week and not able to move forward further because of this issue.

Looking forward your response please.
Thanks
- Deb

**Comment by brainnoise** [comment]
I changed couple of things in the above code and it works for me. I get same eval accuracy both when creating the model and when using saved model. I did not test with jpg image though as I did not have one. 
1. Used the same file name while saving and loading the model.
2. Created model only when saved model is not available. [Just cause model creation is not required in this case. This should not affect the results]
```
if os.path.isfile(filepath):
    model = keras.models.load_model(filepath)
else:
    model = Sequential() 
    ...
```

**Comment by debhazra** [comment]
Hi,

To answer your 1st point  -
**"Used the same file name while saving and loading the model." -**
 By mistake I have given you the different file name while loading the model than the one is being used in training through "save_best_only=True" in the [Checkpointer]. Actually I am using the same file name both the cases. And still I have the problem and therefore I logged this request.

Secondly, my point is not of having any difference in eval accuracy immediately after training or loading the model offline. Both the cases, I got same eval accuracy already. So, here you have the misunderstanding.

I am attaching couple of images for your test. Can you please test and let me know your observation ?
![20180205_174144](https://user-images.githubusercontent.com/28643639/35852658-795c57fa-0b51-11e8-9e99-e69abd601310.jpg)


![20180205_174040](https://user-images.githubusercontent.com/28643639/35852580-4dcc58f6-0b51-11e8-8844-278db84e75bf.jpg)
![20180205_174127](https://user-images.githubusercontent.com/28643639/35852608-60d00c68-0b51-11e8-8fd5-6b7df89769de.jpg)

Thanks
- Deb

**Comment by debhazra** [comment]
Hello,
Any further update please ?

Looking forward your reply please.

Thanks
- Deb

**Comment by debhazra** [comment]
I hope that you got my question. I have the issue that even though I have around 88% evaluation accuracy, but getting same output for different images and never it's right according to the image.

Hope this time I will get your reply soon.

Thanks
- Deb

**Comment by brainnoise** [comment]
I fixed some of the errors in the code snippet, however there are more errors. My guess would be that when you resize the image to 28 x 28, your letter given the proportion might occupy a few undecipherable pixels. Try to compare the image from emnist and your resized image. I verified that this is not a Keras issue. 

If you are looking for quick help with specific implementation, please ask your question on [StackOverflow](https://stackoverflow.com/questions/tagged/keras) or [join the Keras Slack channel](https://keras-slack-autojoin.herokuapp.com/).

**Comment by debhazra** [comment]
Looking forward your reply at least. It it being tested or it's closed from your side. Because my problem remains at the same state.

Thanks
- Deb

**Comment by debhazra** [comment]
I couldn't get your point exactly. Can you please elaborate please ? What error have you fixed in the code snippet as I told you already that I am using the same same .hdf5 file for saving the best model during training as well as at the time of loading the model ? I can see several issues in several sites that keras gives problem during prediction while load the best model from training at different python session. For my case I am trying to understand the probable problem. **Can you please be specific to let me know the problem ? I am more or less sure there is no problem in my code unless it's an issue from Keras.**

Looking forward your reply please.

**Comment by NikhilKothari** [comment]
Hi guy! I'm facing the same issue.

**Comment by lukasijus** [comment]
Hello Guys,

So I don't know if I'm facing the same issue or not, so I will describe my problem below:

I've train my dataset (natural images) on Xception model which gives me >99% and I was amazed. 
But then I started to use model.predict function in many ways and I notices one weird thing. 
When I use to predict images from my test batch everything seem to be OK 99% accuracy. Then I started to capture images from my webcam and what not. Then the predicted results was like dog, dog, dog, dog. I was confused, is it broken? 
So the problem was that the train and test images was already rescaled (from 0:255 -> 0:1) in other words pixels of the images was normalized to 1. 
So when you load your image from the file system or using webcam you have to apply normalization as well. 
Here are code snippet as an example:
`def predict_image(name, image_shape, model, class_names):
    img = cv2.imread(name)
    size = (image_shape[0], image_shape[1])
    img = cv2.resize(img, dsize=size)
    img = np.asarray(img)
    # convert from integers to floats
    img = img.astype('float32')
    # normalize to the range 0-1
    img /= 255.0
    img_to_prediction = np.reshape(img,[1,image_shape[0],image_shape[1],image_shape[2]])
    print(img_to_prediction.shape)
    prediction = model.predict(img_to_prediction)
    print(prediction)
    index = np.argmax(prediction)
    name = class_names[index]
    percent = float("{:.2f}".format(prediction[0][index]))
    print('Predicted: ', name, ' with ', percent, '% accuracy')
    img_name = 'Predicted: ' + name + ' with ' + str(percent) + '% accuracy'
    cv2.imshow(img_name, img)
    cv2.waitKey()`
So the magic lines is this:
`    img = np.asarray(img)
    # convert from integers to floats
    img = img.astype('float32')
    # normalize to the range 0-1
    img /= 255.0`

Hope it will help someone who is facing the same issue.
